File: app/src/main/java/com/example/sergejmitrofanov/librarycatalogue/interactor/BookListRatingSorter.java
package com.example.sergejmitrofanov.librarycatalogue.interactor;

import androidx.annotation.NonNull;
import com.example.sergejmitrofanov.librarycatalogue.entity.Book;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BookListRatingSorter implements BookListSorter {

  @NonNull
  @Override
  public List<Book> sortBooks(@NonNull List<Book> books) {
    List<Book> sortedBooks = new ArrayList<>(books);
    Collections.sort(sortedBooks, (o1, o2) -> {
      int res = o2.rating - o1.rating;
      if (res == 0) {
        return o2.author.compareTo(o1.author);
      }
      return res;
    });
    return sortedBooks;
  }
}

File: app/src/main/java/com/example/sergejmitrofanov/librarycatalogue/interactor/BookListRatingSorter.kt
package com.example.sergejmitrofanov.librarycatalogue.interactor

import com.example.sergejmitrofanov.librarycatalogue.entity.Book

class BookListRatingSorter : BookListSorter {
    override fun sortBooks(books: List<Book>) = books.sortedWith { b1: Book, b2: Book ->
        val result = b2.rating - b1.rating
        if (result == 0) {
            b1.bookTitle.compareTo(b2.bookTitle)
        } else {
            result
        }
    }
}
